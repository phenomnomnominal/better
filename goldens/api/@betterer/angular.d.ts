export declare function angular(configFilePath: string, extraCompilerOptions: CompilerOptions): BettererFileTest;
