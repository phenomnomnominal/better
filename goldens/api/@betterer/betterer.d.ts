export declare function betterer(partialConfig?: BettererStartConfigPartial): Promise<BettererSummary>;
export declare namespace betterer {
    var file: typeof import("./betterer").file;
    var watch: typeof import("./betterer").watch;
}

export declare type BettererBaseConfigPartial = Partial<{
    configPaths: BettererConfigPaths | string;
    cwd: string;
    filters: ReadonlyArray<string | RegExp> | string;
    reporters: BettererConfigReporters;
    resultsPath: string;
    silent: boolean;
    tsconfigPath: string;
}>;

export declare type BettererConfig = {
    allowDiff: boolean;
    configPaths: BettererConfigPaths;
    cwd: string;
    filters: BettererConfigFilters;
    ignores: BettererConfigIgnores;
    reporters: BettererConfigReporters;
    resultsPath: string;
    silent: boolean;
    tsconfigPath: string | null;
    update: boolean;
};

export declare type BettererConfigFilters = ReadonlyArray<RegExp>;

export declare type BettererConfigIgnores = ReadonlyArray<string>;

export declare type BettererConfigPartial = BettererBaseConfigPartial & BettererStartConfigPartial & BettererWatchConfigPartial;

export declare type BettererConfigPaths = ReadonlyArray<string>;

export declare type BettererConfigReporters = ReadonlyArray<BettererConfigReporter>;

export declare type BettererContext = {
    readonly config: BettererConfig;
    readonly lifecycle: Promise<BettererSummaries>;
};

export declare type BettererDeserialise<DeserialisedType extends BettererResultValue, SerialisedType> = (serialised: SerialisedType) => DeserialisedType;

export declare type BettererDiff<DeserialisedType extends BettererResultValue = BettererResultValue, DiffType = null> = {
    expected: DeserialisedType;
    result: DeserialisedType;
    diff: DiffType;
    log: (logger: BettererLogger) => void;
};

export declare type BettererDiffer<DeserialisedType extends BettererResultValue, DiffType> = (expected: DeserialisedType, result: DeserialisedType) => BettererDiff<DeserialisedType, DiffType>;

export declare type BettererFile = BettererFileBase & {
    addIssue(start: number, end: number, message: string, hash?: string): void;
    addIssue(line: number, col: number, length: number, message: string, hash?: string): void;
    addIssue(startLine: number, startCol: number, endLine: number, endCol: number, message: string, hash?: string): void;
};

export declare type BettererFileBase = {
    readonly absolutePath: string;
    readonly hash: string;
    readonly issues: BettererFileIssues;
    readonly key: string;
};

export declare type BettererFileDiff = {
    fixed?: BettererFileIssues;
    existing?: BettererFileIssues;
    new?: BettererFileIssues;
};

export declare type BettererFileGlobs = ReadonlyArray<string | ReadonlyArray<string>>;

export declare type BettererFileIssue = {
    readonly line: number;
    readonly column: number;
    readonly length: number;
    readonly message: string;
    readonly hash: string;
};

export declare type BettererFileIssues = ReadonlyArray<BettererFileIssue>;

export declare type BettererFilePaths = ReadonlyArray<string>;

export declare type BettererFilePatterns = ReadonlyArray<RegExp | ReadonlyArray<RegExp>>;

export declare class BettererFileResolver {
    get cwd(): string;
    constructor(depth?: number);
    excludeΔ(...excludePatterns: BettererFilePatterns): this;
    files(filePaths: BettererFilePaths): Promise<BettererFilePaths>;
    includeΔ(...includePatterns: BettererFileGlobs): this;
    resolve(...pathSegments: Array<string>): string;
    validate(filePaths: BettererFilePaths): Promise<BettererFilePaths>;
}

export declare class BettererFileTest implements BettererTestBase<BettererFileTestResult, BettererFileIssuesMapSerialised, BettererFilesDiff> {
    get config(): BettererTestConfig<BettererFileTestResult, BettererFileIssuesMapSerialised, BettererFilesDiff>;
    readonly isBettererFileTest = "isBettererFileTest";
    readonly isBettererTest = "isBettererTest";
    get isOnly(): boolean;
    get isSkipped(): boolean;
    constructor(_resolver: BettererFileResolver, fileTest: BettererFileTestFunction);
    exclude(...excludePatterns: BettererFilePatterns): this;
    include(...includePatterns: BettererFileGlobs): this;
    only(): this;
    skip(): this;
}

export declare type BettererFileTestDiff = BettererDiff<BettererFileTestResult, BettererFilesDiff>;

export declare type BettererFileTestFunction = (filePaths: BettererFilePaths, fileTestResult: BettererFileTestResult) => MaybeAsync<void>;

export declare type BettererFileTestResult = BettererResultValueComplex & {
    addFile(absolutePath: string, fileText: string): BettererFile;
    getIssues(absolutePath: string): BettererFileIssues;
};

export declare type BettererPrinter<SerialisedType> = (serialised: SerialisedType) => MaybeAsync<string>;

export declare type BettererReporter = {
    configError?(config: BettererConfigPartial, error: BettererError): Promise<void> | void;
    contextStart?(context: BettererContext, lifecycle: Promise<BettererSummaries>): Promise<void> | void;
    contextEnd?(context: BettererContext, summary: BettererSummaries): Promise<void> | void;
    contextError?(context: BettererContext, error: BettererError): Promise<void> | void;
    runsStart?(runs: BettererRuns, files: BettererFilePaths): Promise<void> | void;
    runsEnd?(summary: BettererSummary, files: BettererFilePaths): Promise<void> | void;
    runStart?(run: BettererRun, lifecycle: Promise<void>): Promise<void> | void;
    runEnd?(run: BettererRun): Promise<void> | void;
    runError?(run: BettererRun, error: BettererError): Promise<void> | void;
};

export declare type BettererResult = {
    isNew: boolean;
    value: unknown;
    result: BettererResultValue;
};

export declare type BettererResultValue = BettererResultValueBasic | BettererResultValueComplex;

export declare type BettererResultValueBasic = number;

export declare type BettererResultValueComplex = {
    value: number;
};

export declare type BettererRun = {
    readonly diff: BettererDiff;
    readonly expected: BettererResult;
    readonly filePaths: BettererFilePaths;
    readonly lifecycle: Promise<void>;
    readonly name: string;
    readonly result: BettererResult;
    readonly test: BettererTestConfig;
    readonly timestamp: number;
    readonly isBetter: boolean;
    readonly isComplete: boolean;
    readonly isExpired: boolean;
    readonly isFailed: boolean;
    readonly isNew: boolean;
    readonly isObsolete: boolean;
    readonly isSame: boolean;
    readonly isSkipped: boolean;
    readonly isUpdated: boolean;
    readonly isWorse: boolean;
};

export declare type BettererRunNames = Array<string>;

export declare type BettererRuns = ReadonlyArray<BettererRun>;

export declare type BettererSerialise<DeserialisedType extends BettererResultValue, SerialisedType> = (result: DeserialisedType) => SerialisedType;

export declare type BettererSerialiser<DeserialisedType extends BettererResultValue, SerialisedType = DeserialisedType> = {
    serialise: BettererSerialise<DeserialisedType, SerialisedType>;
    deserialise: BettererDeserialise<DeserialisedType, SerialisedType>;
};

export declare type BettererStartConfigPartial = BettererBaseConfigPartial & Partial<{
    allowDiff: boolean;
    update: boolean;
}>;

export declare type BettererSummaries = Array<BettererSummary>;

export declare type BettererSummary = {
    readonly runs: BettererRuns;
    readonly result: string;
    readonly expected: string | null;
    readonly hasDiff: boolean;
    readonly better: BettererRuns;
    readonly completed: BettererRuns;
    readonly expired: BettererRuns;
    readonly failed: BettererRuns;
    readonly new: BettererRuns;
    readonly obsolete: BettererRuns;
    readonly ran: BettererRuns;
    readonly same: BettererRuns;
    readonly skipped: BettererRuns;
    readonly updated: BettererRuns;
    readonly worse: BettererRuns;
};

export declare class BettererTest<DeserialisedType extends BettererResultValue, SerialisedType, DiffType> implements BettererTestBase<DeserialisedType, SerialisedType, DiffType> {
    get config(): BettererTestConfig<DeserialisedType, SerialisedType, DiffType>;
    readonly isBettererTest = "isBettererTest";
    get isOnly(): boolean;
    get isSkipped(): boolean;
    constructor(config: BettererTestConfigPartial<DeserialisedType, SerialisedType, DiffType>);
    only(): this;
    skip(): this;
}

export declare type BettererTestConfig<DeserialisedType extends BettererResultValue = BettererResultValue, SerialisedType = DeserialisedType, DiffType = null> = {
    constraint: BettererTestConstraint<DeserialisedType>;
    deadline: number;
    goal: BettererTestGoal<DeserialisedType>;
    test: BettererTestFunction<DeserialisedType>;
    differ: BettererDiffer<DeserialisedType, DiffType>;
    printer: BettererPrinter<SerialisedType>;
    serialiser: BettererSerialiser<DeserialisedType, SerialisedType>;
};

export declare type BettererTestConstraint<DeserialisedType extends BettererResultValue> = (result: DeserialisedType, expected: DeserialisedType) => MaybeAsync<BettererConstraintResult>;

export declare type BettererTestFunction<DeserialisedType extends BettererResultValue> = (run: BettererRun) => MaybeAsync<DeserialisedType>;

export declare type BettererTestGoal<DeserialisedType extends BettererResultValue> = (result: DeserialisedType) => MaybeAsync<boolean>;

export declare type BettererWatchConfigPartial = BettererBaseConfigPartial & Partial<{
    ignores: BettererConfigIgnores;
}>;

export declare type BettererWatcher = {
    stop(): Promise<void>;
};

export declare function file(filePath: string, partialConfig?: BettererBaseConfigPartial): Promise<BettererSummary>;

export declare function watch(partialConfig?: BettererWatchConfigPartial): Promise<BettererWatcher>;
