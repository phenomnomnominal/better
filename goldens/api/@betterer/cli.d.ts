export declare type BettererCLIArguments = Array<string>;

export declare type BettererCLICIConfig = {
    config: BettererCLIArguments;
    filter: BettererCLIArguments;
    reporter: BettererCLIArguments;
    results: string;
    silent: boolean;
    tsconfig: string;
};

export declare type BettererCLIInitConfig = {
    config: string;
};

export declare type BettererCLIStartConfig = {
    config: BettererCLIArguments;
    filter: BettererCLIArguments;
    reporter: BettererCLIArguments;
    results: string;
    silent: boolean;
    tsconfig: string;
    update: boolean;
};

export declare type BettererCLIWatchConfig = BettererCLIStartConfig & {
    ignore: BettererCLIArguments;
};

export declare type BettererPackageJSON = {
    version: string;
    scripts: Record<string, string> & {
        betterer: string;
    };
    devDependencies: Record<string, string>;
};

export declare function ciΔ(cwd: string, argv: BettererCLIArguments): Promise<BettererSummary>;

export declare function cliΔ(argv: BettererCLIArguments): void;

export declare function initΔ(cwd: string, argv: BettererCLIArguments): Promise<void>;

export declare function startΔ(cwd: string, argv: BettererCLIArguments): Promise<BettererSummary>;

export declare function watchΔ(cwd: string, argv: BettererCLIArguments): Promise<void>;
