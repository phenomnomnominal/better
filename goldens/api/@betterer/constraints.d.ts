export declare enum BettererConstraintResult {
    better = "better",
    same = "same",
    worse = "worse"
}

export declare function bigger(result: number, expected: number): BettererConstraintResult;

export declare function smaller(result: number, expected: number): BettererConstraintResult;
