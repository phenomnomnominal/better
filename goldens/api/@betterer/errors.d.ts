export declare class BettererError extends Error {
    details: BettererErrorDetails;
    isBettererError: boolean;
    constructor(message: string, ...details: BettererErrorDetails);
}

export declare type BettererErrorDetail = string | Error | BettererError;

export declare type BettererErrorDetails = ReadonlyArray<BettererErrorDetail>;

export declare function isBettererError(err: unknown): err is BettererError;
