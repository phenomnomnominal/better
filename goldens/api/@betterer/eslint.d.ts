export declare function eslint(rules: ESLintRulesConfig): BettererFileTest;
