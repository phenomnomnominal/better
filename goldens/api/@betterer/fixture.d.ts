export declare function createFixtureDirectoryÎ”(fixturesPath: string): Promise<FixtureFactory>;

export declare type Fixture = FixtureFileSystem & {
    logs: ReadonlyArray<string>;
    waitForRun(watcher: BettererWatcher): Promise<BettererSummary>;
    runNames(runs: BettererRuns): BettererRunNames;
};

export declare type FixtureFactory = (fixtureName: string, files: FixtureFileSystemFiles, options?: FixtureOptions) => Promise<Fixture>;

export declare type FixtureFileSystem = {
    paths: Paths;
    deleteDirectory(filePath: string): Promise<void>;
    deleteFile(filePath: string): Promise<void>;
    readFile(filePath: string): Promise<string>;
    resolve(filePath: string): string;
    writeFile(filePath: string, text: string): Promise<void>;
    cleanup(): Promise<void>;
};

export declare type FixtureFileSystemFiles = Record<string, string>;

export declare type FixtureOptions = {
    logFilters?: Array<RegExp>;
};

export declare type Paths = {
    config: string;
    results: string;
    cwd: string;
};
