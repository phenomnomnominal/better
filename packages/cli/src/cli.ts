import commander from 'commander';

import { BettererCLIArguments, BettererPackageJSON } from './types';

enum Command {
  ci = 'ci',
  start = 'start',
  init = 'init',
  watch = 'watch'
}

type CommandName = `${Command}`;

// HACK:
// It's easier to use require than to try to get `await import`
// to work right for the package.json...
/* eslint-disable @typescript-eslint/no-var-requires */
const { version } = require('../package.json') as BettererPackageJSON;

export function cliÎ”(argv: BettererCLIArguments): void {
  commander.version(version);

  commander.command(Command.ci, 'run Betterer in CI mode');
  commander.command(Command.init, 'init Betterer in a project');
  commander.command(Command.start, 'run Betterer');
  commander.command(Command.watch, 'run Betterer in watch mode');

  const args = argv.slice(0);
  const [, , command] = args;
  if (!Command[command as CommandName]) {
    args.splice(2, 0, Command.start);
  }

  commander.parse(args);
}
