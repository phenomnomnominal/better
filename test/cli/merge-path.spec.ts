import { mergeΔ } from '@betterer/cli';

import { createFixture } from '../fixture';

const ARGV = ['node', './bin/betterer'];

describe('betterer cli', () => {
  it('should merge the results file', async () => {
    const { paths, cleanup, readFile } = await createFixture('merge-path', {
      '.betterer.results': `
// BETTERER RESULTS V2.
exports[\`test\`] = {
  value: \`{
|||<<<<<<< our-change
    \\"src/index.ts:315583663\\": [
      [0, 0, 11, \\"TSQuery match\\", \\"3870399096\\"],
      [1, 0, 11, \\"TSQuery match\\", \\"3870399096\\"]
    ]
|||=======
    \\"src/index.ts:913095150\\": [
      [0, 0, 11, \\"TSQuery match\\", \\"3870399096\\"]
    ]
|||>>>>>>> their-change
  }\`
};
      `.replace(/\|||/g, '')
    });

    const resultsPath = paths.results;
    const fixturePath = paths.cwd;

    await mergeΔ(fixturePath, [...ARGV]);

    const merged = await readFile(resultsPath);

    expect(merged).toMatchSnapshot();

    await mergeΔ(fixturePath, [...ARGV, '--results', resultsPath]);

    const unchanged = await readFile(resultsPath);

    expect(unchanged).toMatchSnapshot();

    await cleanup();
  });
});
