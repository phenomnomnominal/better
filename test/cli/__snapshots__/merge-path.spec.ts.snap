// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`betterer cli should merge the results file 1`] = `
"// BETTERER RESULTS V2.
//
// If this file contains merge conflicts, use \`betterer merge\` to automatically resolve them:
// https://phenomnomnominal.github.io/betterer/docs/results-file/#merge
//
exports[\`test\`] = {
  value: \`{
    \\"src/index.ts:913095150\\": [
      [0, 0, 11, \\"TSQuery match\\", \\"3870399096\\"]
    ]
  }\`
};
"
`;

exports[`betterer cli should merge the results file 2`] = `
"// BETTERER RESULTS V2.
//
// If this file contains merge conflicts, use \`betterer merge\` to automatically resolve them:
// https://phenomnomnominal.github.io/betterer/docs/results-file/#merge
//
exports[\`test\`] = {
  value: \`{
    \\"src/index.ts:913095150\\": [
      [0, 0, 11, \\"TSQuery match\\", \\"3870399096\\"]
    ]
  }\`
};
"
`;
